src/transformStateWithClones.js

```diff
@@ -6,8 +6,82 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  // check arguments type
+  if (!isObject(state) && !isObject(actions)) {
+    throw new Error('Invalid arguments');
+  }
+
+  const cloneState = { ...state };
+  const clonedStates = [];
+
+  for (const action of actions) {
+    // get type of action
+    const { type } = action;
+
+    // choice correct action
+    switch (type) {
+      case 'addProperties':
+        const { extraData } = action;
+
+        // get object with adding new properties
+        const addedProperties = addPropertiesToState(cloneState, extraData);
+
+        // add generated object to cloned array
+        clonedStates.push(addedProperties);
```

Comment:
Can we rewrite it so **push** would be performed at the end of each iteration instead of using it at every case.
4. [DONT REPEAT YOURSELF] - If you perform same action in all `switch` cases - do it just once afterwards.